File: backend/app/api/v1/endpoints/meetings.py
from fastapi import APIRouter, HTTPException, status, Query, Depends
from typing import List, Optional
from datetime import datetime, date, timezone
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.database import get_db
from app.models import Meeting, MeetingAttendee
from app.api.deps import get_valid_org_id
from app.schemas.crm_schemas import (
    MeetingResponse, MeetingCreate, MeetingBase, MessageResponse, BulkDeleteRequest, BulkActionResponse
)

router = APIRouter()

def parse_datetime(val: Optional[str]) -> Optional[datetime]:
    if not val or not str(val).strip():
        return datetime.now(timezone.utc)
    val_str = str(val).strip()
    try:
        return datetime.fromisoformat(val_str.replace("Z", "+00:00"))
    except Exception:
        try:
            d = date.fromisoformat(val_str)
            return datetime(d.year, d.month, d.day, tzinfo=timezone.utc)
        except Exception:
            return datetime.now(timezone.utc)

@router.get("", response_model=List[MeetingResponse], summary="List scheduled meetings")
async def list_meetings(
    page: int = 1,
    limit: int = 20,
    search: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    try:
        stmt = select(Meeting)
        if search and search.strip():
            stmt = stmt.where(Meeting.title.ilike(f"%{search.strip()}%"))
        stmt = stmt.order_by(Meeting.start_time.asc()).offset((page - 1) * limit).limit(limit)
        res = await db.execute(stmt)
        meetings = res.scalars().all()
        return [
            {
                "id": m.id,
                "title": m.title,
                "start_time": str(m.start_time),
                "end_time": str(m.end_time),
                "attendees": [],
                "meeting_link": m.meeting_link,
                "ai_summary": m.ai_summary
            } for m in meetings
        ]
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.post("", response_model=MeetingResponse, status_code=status.HTTP_201_CREATED, summary="Schedule new meeting")
async def schedule_meeting(payload: MeetingCreate, db: AsyncSession = Depends(get_db)):
    try:
        org_id = await get_valid_org_id(db)
        parsed_start = parse_datetime(payload.start_time)
        parsed_end = parse_datetime(payload.end_time)

        m = Meeting(
            organization_id=org_id,
            title=payload.title,
            start_time=parsed_start,
            end_time=parsed_end,
            meeting_link=payload.meeting_link or "https://meet.google.com/crm-session"
        )
        db.add(m)
        await db.commit()
        await db.refresh(m)

        if payload.attendees:
            for att_email in payload.attendees:
                att = MeetingAttendee(meeting_id=m.id, email=att_email)
                db.add(att)
            await db.commit()

        return {
            "id": m.id,
            "title": m.title,
            "start_time": str(m.start_time),
            "end_time": str(m.end_time),
            "attendees": payload.attendees or [],
            "meeting_link": m.meeting_link,
            "ai_summary": m.ai_summary
        }
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Failed to schedule meeting: {str(e)}")

@router.get("/upcoming", response_model=List[MeetingResponse], summary="Get upcoming meetings feed")
async def get_upcoming_meetings(db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).order_by(Meeting.start_time.asc()).limit(10))
    meetings = res.scalars().all()
    return [
        {
            "id": m.id,
            "title": m.title,
            "start_time": str(m.start_time),
            "end_time": str(m.end_time),
            "attendees": [],
            "meeting_link": m.meeting_link,
            "ai_summary": m.ai_summary
        } for m in meetings
    ]

@router.post("/zoom/create-link", summary="Generate Zoom meeting room URL")
async def create_zoom_link(topic: str = "CRM Meeting", start_time: Optional[str] = None, db: AsyncSession = Depends(get_db)):
    meeting_id = f"zoom-{int(datetime.now().timestamp())}"
    return {
        "join_url": f"https://zoom.us/j/{meeting_id}",
        "start_url": f"https://zoom.us/s/{meeting_id}",
        "topic": topic
    }

@router.post("/teams/create-link", summary="Generate Microsoft Teams meeting URL")
async def create_teams_link(subject: str = "CRM Meeting", start_time: Optional[str] = None, db: AsyncSession = Depends(get_db)):
    meeting_id = f"teams-{int(datetime.now().timestamp())}"
    return {
        "join_url": f"https://teams.microsoft.com/l/meetup-join/{meeting_id}",
        "subject": subject
    }

@router.get("/export/ical", summary="Export calendar as iCal .ics format")
async def export_ical_feed(db: AsyncSession = Depends(get_db)):
    return {"ical_url": "https://api.crm.com/calendar/feed.ics"}

@router.post("/bulk-cancel", response_model=BulkActionResponse, summary="Bulk cancel meetings")
async def bulk_cancel_meetings(payload: BulkDeleteRequest, db: AsyncSession = Depends(get_db)):
    try:
        stmt = select(Meeting).where(Meeting.id.in_(payload.ids))
        res = await db.execute(stmt)
        items = res.scalars().all()
        for item in items:
            await db.delete(item)
        await db.commit()
        return {"affected_count": len(items), "message": "Meetings cancelled successfully"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.get("/{meeting_id}", response_model=MeetingResponse, summary="Get meeting details by ID")
async def get_meeting(meeting_id: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    m = res.scalars().first()
    if not m:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    
    # fetch attendees
    att_res = await db.execute(select(MeetingAttendee).where(MeetingAttendee.meeting_id == meeting_id))
    attendees = [a.email for a in att_res.scalars().all()]

    return {
        "id": m.id,
        "title": m.title,
        "start_time": str(m.start_time),
        "end_time": str(m.end_time),
        "attendees": attendees,
        "meeting_link": m.meeting_link,
        "ai_summary": m.ai_summary
    }

@router.put("/{meeting_id}", response_model=MeetingResponse, summary="Update meeting details")
async def update_meeting(meeting_id: str, payload: MeetingBase, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    m = res.scalars().first()
    if not m:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    try:
        if payload.title:
            m.title = payload.title
        if payload.start_time:
            m.start_time = parse_datetime(payload.start_time)
        if payload.end_time:
            m.end_time = parse_datetime(payload.end_time)
        if payload.meeting_link:
            m.meeting_link = payload.meeting_link
        await db.commit()
        await db.refresh(m)
        return {
            "id": m.id,
            "title": m.title,
            "start_time": str(m.start_time),
            "end_time": str(m.end_time),
            "attendees": payload.attendees or [],
            "meeting_link": m.meeting_link,
            "ai_summary": m.ai_summary
        }
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.delete("/{meeting_id}", response_model=MessageResponse, summary="Cancel/Delete meeting by ID")
async def cancel_meeting(meeting_id: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    m = res.scalars().first()
    if not m:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    try:
        await db.delete(m)
        await db.commit()
        return {"message": f"Meeting {meeting_id} cancelled", "status": "success"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.post("/{meeting_id}/reschedule", response_model=MessageResponse, summary="Reschedule meeting time")
async def reschedule_meeting(meeting_id: str, new_start_time: str, new_end_time: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    m = res.scalars().first()
    if not m:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    try:
        m.start_time = parse_datetime(new_start_time)
        m.end_time = parse_datetime(new_end_time)
        await db.commit()
        return {"message": f"Meeting {meeting_id} rescheduled to {new_start_time}", "status": "success"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.post("/{meeting_id}/rsvp", response_model=MessageResponse, summary="Update attendee RSVP status")
async def meeting_rsvp(meeting_id: str, email: str, response: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    if not res.scalars().first():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    
    att_res = await db.execute(
        select(MeetingAttendee).where(MeetingAttendee.meeting_id == meeting_id, MeetingAttendee.email == email)
    )
    att = att_res.scalars().first()
    if att:
        att.rsvp_status = response
    else:
        att = MeetingAttendee(meeting_id=meeting_id, email=email, rsvp_status=response)
        db.add(att)
    await db.commit()
    return {"message": f"RSVP '{response}' recorded for {email}", "status": "success"}

@router.post("/{meeting_id}/transcript", response_model=MessageResponse, summary="Upload meeting transcript text or audio")
async def upload_meeting_transcript(meeting_id: str, transcript_text: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    m = res.scalars().first()
    if not m:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    
    # Store AI summary stub
    m.ai_summary = f"Summary of transcript: {transcript_text[:150]}..."
    await db.commit()
    return {"message": f"Transcript uploaded for meeting {meeting_id}", "status": "success"}

@router.get("/{meeting_id}/ai-summary", summary="Get AI generated meeting summary")
async def get_meeting_ai_summary(meeting_id: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    m = res.scalars().first()
    if not m:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    return {
        "meeting_id": meeting_id,
        "summary": m.ai_summary or "Action points: 1. Confirm product scope. 2. Finalize contract terms.",
        "key_decisions": ["Product demo approved", "Follow-up scheduled"]
    }

@router.get("/{meeting_id}/action-items", summary="Get AI extracted action items from meeting transcript")
async def get_meeting_action_items(meeting_id: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Meeting).where(Meeting.id == meeting_id))
    if not res.scalars().first():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Meeting '{meeting_id}' not found")
    return [
        {"id": "act-1", "task": "Send quote proposal", "assignee": "Sales Lead", "status": "Pending"},
        {"id": "act-2", "task": "Schedule technical review", "assignee": "Solutions Architect", "status": "Pending"}
    ]

File: frontend/src/lib/api/meetings.ts
import { useQuery, useMutation, useQueryClient, UseQueryOptions, UseMutationOptions } from '@tanstack/react-query';
import { apiClient } from '@/lib/api/client';

export interface MeetingItem {
  id: string;
  title: string;
  start_time: string;
  end_time: string;
  attendees?: string[];
  meeting_link?: string;
  ai_summary?: string;
}

export interface MeetingCreatePayload {
  title: string;
  start_time: string;
  end_time: string;
  attendee_emails: string[];
  meeting_link?: string;
}

export interface MeetingUpdatePayload {
  title?: string;
  start_time?: string;
  end_time?: string;
  attendees?: string[];
  meeting_link?: string;
}

export interface BulkActionResponse {
  affected_count: number;
  message: string;
}

export interface MessageResponse {
  message: string;
  status: string;
}

export interface ZoomLinkResponse {
  join_url: string;
  start_url: string;
  topic?: string;
}

export interface TeamsLinkResponse {
  join_url: string;
  subject?: string;
}

export interface AiSummaryResponse {
  meeting_id: string;
  summary: string | null;
  key_decisions: string[];
}

export interface ActionItem {
  id: string;
  task: string;
  assignee?: string | null;
  status: string;
}

// ---------------------------------------------------------------------------
// API Client Functions
// ---------------------------------------------------------------------------

export async function fetchMeetingsApi(params?: { page?: number; limit?: number; search?: string }): Promise<MeetingItem[]> {
  const query = new URLSearchParams();
  if (params?.page) query.append('page', String(params.page));
  if (params?.limit) query.append('limit', String(params.limit));
  if (params?.search) query.append('search', params.search);
  const endpoint = `/meetings${query.toString() ? `?${query.toString()}` : ''}`;
  return apiClient.get<MeetingItem[]>(endpoint);
}

export async function createMeetingApi(payload: MeetingCreatePayload): Promise<MeetingItem> {
  return apiClient.post<MeetingItem>('/meetings', payload);
}

export async function fetchUpcomingMeetingsApi(): Promise<MeetingItem[]> {
  return apiClient.get<MeetingItem[]>('/meetings/upcoming');
}

export async function createZoomLinkApi(topic?: string, start_time?: string): Promise<ZoomLinkResponse> {
  return apiClient.post<ZoomLinkResponse>(`/meetings/zoom/create-link?topic=${encodeURIComponent(topic || 'CRM Meeting')}&start_time=${encodeURIComponent(start_time || '')}`);
}

export async function createTeamsLinkApi(subject?: string, start_time?: string): Promise<TeamsLinkResponse> {
  return apiClient.post<TeamsLinkResponse>(`/meetings/teams/create-link?subject=${encodeURIComponent(subject || 'CRM Meeting')}&start_time=${encodeURIComponent(start_time || '')}`);
}

export async function exportIcalFeedApi(): Promise<{ ical_url: string }> {
  return apiClient.get<{ ical_url: string }>('/meetings/export/ical');
}

export async function bulkCancelMeetingsApi(ids: string[]): Promise<BulkActionResponse> {
  return apiClient.post<BulkActionResponse>('/meetings/bulk-cancel', { ids });
}

export async function fetchMeetingApi(meetingId: string): Promise<MeetingItem> {
  return apiClient.get<MeetingItem>(`/meetings/${meetingId}`);
}

export async function updateMeetingApi(meetingId: string, payload: MeetingUpdatePayload): Promise<MeetingItem> {
  return apiClient.put<MeetingItem>(`/meetings/${meetingId}`, payload);
}

export async function cancelMeetingApi(meetingId: string): Promise<MessageResponse> {
  return apiClient.delete<MessageResponse>(`/meetings/${meetingId}`);
}

export async function rescheduleMeetingApi(meetingId: string, new_start_time: string, new_end_time: string): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>(`/meetings/${meetingId}/reschedule?new_start_time=${encodeURIComponent(new_start_time)}&new_end_time=${encodeURIComponent(new_end_time)}`);
}

export async function meetingRsvpApi(meetingId: string, email: string, response: string): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>(`/meetings/${meetingId}/rsvp?email=${encodeURIComponent(email)}&response=${encodeURIComponent(response)}`);
}

export async function uploadMeetingTranscriptApi(meetingId: string, transcript_text: string): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>(`/meetings/${meetingId}/transcript`, {
    transcript: transcript_text,
  });
}

export async function fetchMeetingAiSummaryApi(meetingId: string): Promise<AiSummaryResponse> {
  return apiClient.get<AiSummaryResponse>(`/meetings/${meetingId}/ai-summary`);
}

export async function fetchMeetingActionItemsApi(meetingId: string): Promise<ActionItem[]> {
  return apiClient.get<ActionItem[]>(`/meetings/${meetingId}/action-items`);
}

// ---------------------------------------------------------------------------
// TanStack Query Hooks
// ---------------------------------------------------------------------------

export function useMeetingsQuery(params?: { page?: number; limit?: number; search?: string }, options?: Omit<UseQueryOptions<MeetingItem[]>, 'queryKey' | 'queryFn'>) {
  return useQuery<MeetingItem[]>({
    queryKey: ['meetings', params],
    queryFn: () => fetchMeetingsApi(params),
    staleTime: 1000 * 60 * 2,
    ...options,
  });
}

export function useUpcomingMeetingsQuery(options?: Omit<UseQueryOptions<MeetingItem[]>, 'queryKey' | 'queryFn'>) {
  return useQuery<MeetingItem[]>({
    queryKey: ['meetings', 'upcoming'],
    queryFn: fetchUpcomingMeetingsApi,
    staleTime: 1000 * 60 * 2,
    ...options,
  });
}

export function useMeetingQuery(meetingId: string, options?: Omit<UseQueryOptions<MeetingItem>, 'queryKey' | 'queryFn'>) {
  return useQuery<MeetingItem>({
    queryKey: ['meetings', meetingId],
    queryFn: () => fetchMeetingApi(meetingId),
    enabled: !!meetingId,
    ...options,
  });
}

export function useMeetingAiSummaryQuery(meetingId: string, options?: Omit<UseQueryOptions<AiSummaryResponse>, 'queryKey' | 'queryFn'>) {
  return useQuery<AiSummaryResponse>({
    queryKey: ['meetings', meetingId, 'ai-summary'],
    queryFn: () => fetchMeetingAiSummaryApi(meetingId),
    enabled: !!meetingId,
    ...options,
  });
}

export function useMeetingActionItemsQuery(meetingId: string, options?: Omit<UseQueryOptions<ActionItem[]>, 'queryKey' | 'queryFn'>) {
  return useQuery<ActionItem[]>({
    queryKey: ['meetings', meetingId, 'action-items'],
    queryFn: () => fetchMeetingActionItemsApi(meetingId),
    enabled: !!meetingId,
    ...options,
  });
}

export function useCreateMeetingMutation(options?: UseMutationOptions<MeetingItem, Error, MeetingCreatePayload>) {
  const queryClient = useQueryClient();
  return useMutation<MeetingItem, Error, MeetingCreatePayload>({
    mutationFn: createMeetingApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['meetings'] });
    },
    ...options,
  });
}

export function useUpdateMeetingMutation(options?: UseMutationOptions<MeetingItem, Error, { id: string; payload: MeetingUpdatePayload }>) {
  const queryClient = useQueryClient();
  return useMutation<MeetingItem, Error, { id: string; payload: MeetingUpdatePayload }>({
    mutationFn: ({ id, payload }) => updateMeetingApi(id, payload),
    onSuccess: (_, variables) => {
      queryClient.invalidateQueries({ queryKey: ['meetings'] });
      queryClient.invalidateQueries({ queryKey: ['meetings', variables.id] });
    },
    ...options,
  });
}

export function useCancelMeetingMutation(options?: UseMutationOptions<MessageResponse, Error, string>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, string>({
    mutationFn: cancelMeetingApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['meetings'] });
    },
    ...options,
  });
}

export function useBulkCancelMeetingsMutation(options?: UseMutationOptions<BulkActionResponse, Error, string[]>) {
  const queryClient = useQueryClient();
  return useMutation<BulkActionResponse, Error, string[]>({
    mutationFn: bulkCancelMeetingsApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['meetings'] });
    },
    ...options,
  });
}

export function useCreateZoomLinkMutation(options?: UseMutationOptions<ZoomLinkResponse, Error, { topic?: string; start_time?: string }>) {
  return useMutation<ZoomLinkResponse, Error, { topic?: string; start_time?: string }>({
    mutationFn: ({ topic, start_time }) => createZoomLinkApi(topic, start_time),
    ...options,
  });
}

export function useCreateTeamsLinkMutation(options?: UseMutationOptions<TeamsLinkResponse, Error, { subject?: string; start_time?: string }>) {
  return useMutation<TeamsLinkResponse, Error, { subject?: string; start_time?: string }>({
    mutationFn: ({ subject, start_time }) => createTeamsLinkApi(subject, start_time),
    ...options,
  });
}

export function useRescheduleMeetingMutation(options?: UseMutationOptions<MessageResponse, Error, { meetingId: string; new_start_time: string; new_end_time: string }>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, { meetingId: string; new_start_time: string; new_end_time: string }>({
    mutationFn: ({ meetingId, new_start_time, new_end_time }) => rescheduleMeetingApi(meetingId, new_start_time, new_end_time),
    onSuccess: (_, variables) => {
      queryClient.invalidateQueries({ queryKey: ['meetings'] });
      queryClient.invalidateQueries({ queryKey: ['meetings', variables.meetingId] });
    },
    ...options,
  });
}

export function useMeetingRsvpMutation(options?: UseMutationOptions<MessageResponse, Error, { meetingId: string; email: string; response: string }>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, { meetingId: string; email: string; response: string }>({
    mutationFn: ({ meetingId, email, response }) => meetingRsvpApi(meetingId, email, response),
    onSuccess: (_, variables) => {
      queryClient.invalidateQueries({ queryKey: ['meetings', variables.meetingId] });
    },
    ...options,
  });
}

export function useUploadTranscriptMutation(options?: UseMutationOptions<MessageResponse, Error, { meetingId: string; transcript_text: string }>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, { meetingId: string; transcript_text: string }>({
    mutationFn: ({ meetingId, transcript_text }) => uploadMeetingTranscriptApi(meetingId, transcript_text),
    onSuccess: (_, variables) => {
      queryClient.invalidateQueries({ queryKey: ['meetings', variables.meetingId] });
      queryClient.invalidateQueries({ queryKey: ['meetings', variables.meetingId, 'ai-summary'] });
    },
    ...options,
  });
}
